### ISSUE TITLE
Question: how to stop a server gracefully?

### ISSUE THREAD

**Issue by rogierc** [eclipse/californium#1727]
[rfc7641](https://datatracker.ietf.org/doc/html/rfc7641#section-3.2) states that an error notification should be sent to observing clients when the resource changes in a way that would cause a  normal GET request at that time to return a non-2.xx response.

Californium complies by sending an 4.04 to observing clients when a resource is deleted. But when a server is stopped, it does not. Because a GET on the resource after stopping would fail, it would be according to the idea of the rfc to send 4.04 to observers, when going down gracefully. 

One could explicitly delete the resources before calling server.stop(). But the endpoint will be stopped also and will not proces  messages. The 4.04 message wil not be sent out without waiting time between the resource.delete() and server.stop().

Waiting time is not elegant and will delay shutting down unnecessarily. Is there another way to get the error-message sent?

**Comment by boaks** [comment]
That is not implemented.

I'm not sure, if it is a good idea. At least not for all cases. Assume, that there are many clients, that may cause a lot of traffic. Using a CON response to ensure would also cause the server to stay until that times out or the client ACKs, which also not that sure. Using NON response may get dropped. Many things to consider ...

**Comment by boaks** [comment]
Maybe a server-side function to "stop" the observation without delete the resource? With callbacks, when either the message was sent (NON) or the ACK was received (CON)?

**Comment by boaks** [comment]
:-)

The method is already there:

`CoapResource.clearAndNotifyObserveRelations(ResponseCode code)`

With 

`CoapResource.setObserveType(Type type)`

you may check, what works better for you.

**Comment by rogierc** [comment]
> At least not for all cases

Agree. Although you could interpret rfc7641 in the sense that sending an error message is an appropriate thing to do when a server is going down.

In some use cases it´s important that an observing client soon discovers that the server (thus the resource) has gone away. Setting the Max-Age to a small number is not attractive because that adds a lot of traffic for covering a situation that rarely occurs. 

> The method is already there

Yes, the server can call `clearAndNotifyObserveRelations`, in fact `CoapResource.delete()` does that. When going down the server could do something like

```
for ( CoapResouce resource : resources ){  resource.delete(); } 
server.stop();
```

But this doesn't work, because the server stops the endpoint immediately and the 4.04 messages don't get sent.
Adding sleep works:
```
for ( CoapResouce resource : resources ){  resource.delete(); } 
Thread.sleep( WAIT_FOR_SOME_TIME );
server.stop();
```
but is a crude solution. To get a gracefull shutdown one would like to be able to do something like
```
for ( CoapResouce resource : resources ){  resource.delete(); } 
server.finishExchanges();
server.stop();
```
where finishExchanges executes a kind of preparation fase for going down and proceses existing exhanges. The method returns when all exchanges have finished.

**Comment by boaks** [comment]
Let me ask:

- how often do you stop the server?
 
- is there more needed? In LwM2M (leshan) it was asked recently to make the sever-side observe persistent and possible to restart.

In my own experience, restarts are rare. Last year there was a discussion about a graceful restart for DTLS. That's now implemented. I received close to no feedback for it. I don't plan to use observe/notify, so at least for now, I don't see, that I will implement and test these observe/notify extensions. If someone, who uses it, has interest to contribute it, feel free.

**Comment by rogierc** [comment]
> how often do you stop the server?

Not often. Reason for going down is maintenance or the occasional malfunction of some kind. Maybe once a month or more likely once a year. The server is expected to be available 24x7. Outage needs to be minimised. Therefore it's important for clients to detect the observe is broken and to revive it as soon as possible. 

> In LwM2M (leshan) it was asked recently to make the sever-side observe persistent and possible to restart.

Such a feature would certainly help. When the server does a restart fast enough, clients wouldn't even notice it has.

> Last year there was a discussion about a graceful restart for DTLS.

We don't use it now, but probably will in the future (for what it's worth ;-))
Would parts of this implementation be reusable for oberve relation persistence, maybe?

**Comment by boaks** [comment]
@rogierc 

I don't think, anyone is working on this. Should it be kept open or could it be closed?

**Comment by rogierc** [comment]
>Should it be kept open or could it be closed?

OK to close.

A workable solution for a gracefull shutdown is to delete the resources, which results in sending 4.04 notifications to observers. Apart from issues mentioned in #1754 will the observers be aware that the resources are gone. Important is to keep the endpoint alive until all notification exchanges have completed. It would be nice to have a way to establish that all notifications are really complete before stopping the endpoint (suggestions?), but a grace period will do in many use cases probably.

**Comment by boaks** [comment]
The point, longer ago, with that coap graceful shutdown was, that with the coap-role exchange of lwm2m it's hard to distinguish from messages, which should be rejected, and those, which are considered to be processed as part of the graceful shutdown.

So, your endpoint should know, which messages should be denied, and which should be processed. From both APIs north (CoAP application) or south (Connector). That may includes incoming and outgoing exchanges. Once you clear about that, you may add the APIs to the endpoint to be able to wait on the "pending exchanges to have finished". That may take sometimes longer than expected, especially, if some of the other peers have disappeared and there will not send the expected message back.

**Comment by boaks** [comment]
With the experience of the last years:
I guess no-one has the  interest to work on this.
I can leave it open, but I prefer to close it.

**Comment by rogierc** [comment]
> The point, longer ago, with that coap graceful shutdown was, that with the coap-role exchange of lwm2m it's hard to distinguish from messages, which should be rejected, and those, which are considered to be processed as part of the graceful shutdown.

The right way to do a gracefull shutdown seems to be application dependend, then. To implement gracefull shutdown the application should be able to 'see' what exchanges are incomplete and decide to linger or go down. What means are available right now to do that? Using MessageInterceptor for this is not practical.

**Comment by boaks** [comment]
> to 'see' what exchanges are incomplete

That's one part. 
The other one is to distinguish, which messages are still intended to be processed, and which are not and should be dropped. Therefore a `MessageInterceptor` should work. 

Though the exchanges are kept in a `MessageExchangeStore`, you may start using a custom one, provided to the `CoapEndpoint.Builder`. That implementation may be  based on a customized `InMemoryMessageExchangeStore`. If you have results, we may also consider to adapt the `InMemoryMessageExchangeStore` to make such adaption easier.

**Comment by rogierc** [comment]
Ok. This gives a direction to solving the issue. Thanks!!

We have a workaround that works for many cases (just stop the endpoint not too fast), so need to see when to come around to customising `MessageExchangeStore`. For my part no need to keep the issue open - the question has been answered :-)

**Comment by boaks** [comment]
Any outcome so far?
Do you see a API, which should be extended?
just to mention:
Implementing a "serialization" of the server observe side is not that easy, but the effect may be really nice.
So, it mainly need volunteers. 
At least the dtls graceful restart makes in my opinion the usage much easier. Therefore I started to extend that API for easier integration, see [EncryptedServersSerializationUtil](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/server/EncryptedServersSerializationUtil.java) .

**Comment by boaks** [comment]
@rogierc 

Any outcome in the meantime?

Just to mention:
I redesigned the persistence API, see [PersistentComponent](https://github.com/eclipse/californium/blob/master/element-connector/src/main/java/org/eclipse/californium/elements/PersistentComponent.java) and [EncryptedPersistentComponentUtil](https://github.com/eclipse/californium/blob/master/element-connector/src/main/java/org/eclipse/californium/elements/util/EncryptedPersistentComponentUtil.java). At least the intention was, to support persistence for other layer's state as well.

**Comment by boaks** [comment]
With #2014 the new data-model of the server-side's observe may be persisted and reloaded. As long as the server is stopped to be started fast again, that would even prevent the clients from being affected.

Beside of the initiating observe-request, the `ObserveNotificationOrderer` of the `CoapResource` must also be persisted. The `interestCheckTimer` and `interestCheckCounter` maybe persisted or just restarted. Also a currently pending CON-Notification may be persisted, or alternatively, a CON Notification may be used always on restart (I think, that's the easier and my preferred  way).
